keras/layers/normalization/spectral_normalization.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2023 The Keras Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+import tensorflow.compat.v2 as tf
+
+from keras.layers.rnn import Wrapper
+
+# isort: off
+from tensorflow.python.util.tf_export import keras_export
+
+
+# Adapted from TF-Addons implementation
+@keras_export("keras.layers.SpectralNormalization", v1=[])
+class SpectralNormalization(Wrapper):
+    """Performs spectral normalization on weights.
+    This wrapper controls the Lipschitz constant of the layer by
+    constraining its spectral norm, which can stabilize the training of GANs.
+    See [Spectral Normalization for GAN](https://arxiv.org/abs/1802.05957).
+    Wrap `tf.keras.layers.Conv2D`:
+    >>> x = np.random.rand(1, 10, 10, 1)
+    >>> conv2d = SpectralNormalization(tf.keras.layers.Conv2D(2, 2))
+    >>> y = conv2d(x)
+    >>> y.shape
+    TensorShape([1, 9, 9, 2])
+    Wrap `tf.keras.layers.Dense`:
+    >>> x = np.random.rand(1, 10, 10, 1)
+    >>> dense = SpectralNormalization(tf.keras.layers.Dense(10))
+    >>> y = dense(x)
+    >>> y.shape
+    TensorShape([1, 10, 10, 10])
+    Args:
+      layer: A `tf.keras.layers.Layer` instance that
+        has either `kernel` or `embeddings` attribute.
```

Comment:
"a"

Reply:
"A" to "a"?

Reply:
"a `kernel` or an `embeddings` attribute"